dependencies.yaml

```
@@ -377,6 +378,33 @@ dependencies:
               - *libcusolver114
               - *libcusparse_dev114
               - *libcusparse114
+  cuda_wheels:
+    specific:
+      - output_types: pyproject
+        matrices:
+          - matrix:
+              cuda: "12.*"
+              cuda_suffixed: "true"
+              use_cuda_wheels: "true"
+            packages:
+              - nvidia-cublas-cu12
+              - nvidia-curand-cu12
+              - nvidia-cusolver-cu12
+              - nvidia-cusparse-cu12
+          - matrix:
+              cuda: "11.*"
+              cuda_suffixed: "true"
+              use_cuda_wheels: "true"
+            packages:
+          - matrix:
+              use_cuda_wheels: "false"
+            packages:
+          - matrix:
+            packages:
+              - nvidia-cublas
+              - nvidia-curand
+              - nvidia-cusolver
+              - nvidia-cusparse
```

```suggestion
          # if use_cuda_wheels=false is provided, do not add dependencies on any CUDA wheels
          # (e.g. for DLFW and pip devcontainers)
          - matrix:
              use_cuda_wheels: "false"
            packages:
          # if no matching matrix selectors passed, list the unsuffixed packages
          # (just as a source of documentation, as this populates pyproject.toml in source control)
          - matrix:
            packages: *cuda_wheels_cu12
```

This is sufficiently subtle that I really think it deserves code comments. Are my suggested comments here a correct summary of the reason for having two fallback matrices?

I also think we should just pick one of the lists above instead of repeating all of these. And that it should be the CUDA 12 list, to help with https://github.com/rapidsai/build-planning/issues/68

---

> Are my suggested comments here a correct summary of the reason for having two fallback matrices?

Yes, looks right to me.

> I also think we should just pick one of the lists above instead of repeating all of these.

I don't agree. The "default" list in `pyproject.toml` has unsuffixed packages for all of the other dependencies, even though those wheels don't exist. Why should the CUDA wheels have a suffix by default?

---

I SWEAR that when I was looking at this diff, the CUDA 11 list had all the unsuffixed names in it as well. And that that helped me to understand that unlike with RAPIDS libraries, the unsuffixed names are real distribution names installable from `pypi.nvidia.com`. For example, https://pypi.nvidia.com/nvidia-cublas/.

So I thought "if these are going to be real distribution names anyway, we might as well prefer the CUDA 12 ones".

But I don't see those in the CUDA 11 list in `dependencies.yaml` any more (and I understand why, after re-reading `CMakeLists.txt`), so I guess that point no longer holds, and it's fine to leave the unsuffixed names in `pyproject.toml` in source control.